apps/extension/src/core/media/model.ts

```diff
@@ -0,0 +1,239 @@
+import fetchAdapter from "@vespaiach/axios-fetch-adapter"
+import type { AxiosInstance, AxiosRequestConfig } from "axios"
+import axios, { AxiosError } from "axios"
+import axiosRetry, { exponentialDelay } from "axios-retry"
+import objectHash from "object-hash"
+import { type ChatMessage, ErrorCode, type MediaOutput, MediaMimeType, type ThreeDOptions } from "window.ai"
+
+import { type Err, type Result, err, ok } from "~core/utils/result-monad"
+import { definedValues, parseDataChunks } from "~core/utils/utils"
+
+// These options are specific to the model shape and archetype
+export interface MediaModelConfig {
+  defaultBaseUrl: string
+  identifier: string
+  getPath: (request: RequestData) => string
+  transformForRequest: (
+    request: RequestData,
+    meta: RequestMetadata
+  ) => Record<string, unknown>
+  transformResponse: (res: unknown) => MediaOutput[]
+
+  // Optionals
+  getRoutePath?: (request: RequestData) => string | null
+  overrideModelParam?: (request: RequestData) => string | null
+  customHeaders?: Record<string, string>
+  authPrefix?: string
+  debug?: boolean
+  retries?: number
+  cacheGet?: CacheGetter
+  cacheSet?: CacheSetter
+  adapter?: AxiosRequestConfig["adapter"]
+}
+
+export interface RequestOptions {
+  baseUrl?: string
+  apiKey?: string | null
+  model?: string | null
+  origin?: string | null
+  num_generations?: number
+  mimeType?: MediaMimeType | null
+  num_inference_steps?: number | null
+  timeout?: number
+  user_identifier?: string | null
+  adapter?: AxiosRequestConfig["adapter"] | null
+}
+
+type RequestPrompt = { prompt: string; }
+
+// export interface RequestPrompt extends Partial<RequestPromptBasic>
+
+export type RequestData = Omit<
+  Required<RequestOptions>,
+  "user_identifier" | "timeout" | "apiKey" | "origin" | "adapter" // These do not affect output of the model
+> &
+  Pick<Required<MediaModelConfig>, "identifier"> & // To distinguish btw providers with same-name models
+  RequestPrompt
+
+export type RequestMetadata = Pick<RequestOptions, "user_identifier">
+
+// TODO cache statistics and log probs etc
+export type CacheGetter = (id: string) => Promise<MediaOutput[] | null | undefined>
+
+export type CacheSetter = (data: {
+  id: string
+  prompt: RequestData
+  completion: MediaOutput[]
+}) => Promise<unknown>
+
+export class MediaModel {
+  public api: AxiosInstance
+  public config: Required<MediaModelConfig>
+  public defaultOptions: Required<RequestOptions>
+  
+  constructor(config: MediaModelConfig, opts: RequestOptions = {}) {
+    // Defaults
+    this.config = this.addDefaults(config)
+    this.defaultOptions = {
+      baseUrl: this.config.defaultBaseUrl,
+      model: null,
+      origin: null,
+      apiKey: null,
+      timeout: 60000,
+      user_identifier: null,
+      num_generations: 1,
+      num_inference_steps: null,
```

Comment:
i would have expected the default, `32`, to appear here, rather than that other file

Reply:
changed, but new models will need to define their own default num_inference_steps in their model constructor if they want something other than 32 as the default